src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs

```
@@ -761,106 +766,108 @@ public int Seek(ref Vector<byte> byte0Vector, ref Vector<byte> byte1Vector, ref
-                    if (byteEquals[offset + 2] != 0) return offset + 2;
-                    return offset + 3;
-                }
-                if (byteEquals[offset + 4] != 0) return offset + 4;
-                if (byteEquals[offset + 5] != 0) return offset + 5;
-                if (byteEquals[offset + 6] != 0) return offset + 6;
-                return offset + 7;
-            }
-            throw new InvalidOperationException();
+            // Flag least significant power of two bit
+            var powerOfTwoFlag = (ulong)(longValue ^ (longValue - 1));
+            // Shift all powers of two into the high byte and extract
+            var foundByteIndex = (int)((powerOfTwoFlag * _xorPowerOfTwoToHighByte) >> 57);
+            // Single LEA instruction with jitted const (using function result)
+            return i * 8 + foundByteIndex;
         }
 
         /// <summary>
         /// Save the data at the current location then move to the next available space.
         /// </summary>
         /// <param name="data">The byte to be saved.</param>
         /// <returns>true if the operation successes. false if can't find available space.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public bool Put(byte data)
         {
-            if (_block == null)
+            var block = _block;
+            if (block == null)
             {
                 return false;
             }
 
-            var block = _block;
             var index = _index;
-            while (true)
+
+            // Always set wasLastBlock before checking .End to avoid race which may cause data loss
+            var wasLastBlock = block.Next == null;
+            if (index < block.End)
             {
+                _index = index + 1;
+                block.Array[index] = data;
+                return true;
+            }
+
+            return wasLastBlock ? false : PutMultiBlock(data);
```

Similar to `Skip()`, the return value of this is never checked in non-test code. It's probably better to throw instead of return false.